Added doc for Ztunnel inpod mode + updated ztunnel L4 guide for 1.21 function (#14611)
istio/istio.io: content/en/docs/ops/ambient/usage/ztunnel/index.md

## Current Caveats {#caveats}

> Ztunnel proxies are automatically installed when one of the supported installation methods is used to install Istio ambient mesh. The minimum Istio version required for Istio ambient mode is `1.21.0` (note: although Istio Ambient is functional since version 1.18.0, it is strongly recommended at this time to use version 1.21.0 or newer). In general Istio in ambient mode supports the existing Istio APIs that are supported in sidecar proxy mode. Since the ambient functionality is currently at an alpha release level, the following is a list of feature restrictions or caveats in the current release of Istio's ambient functionality (as of the `1.21.0` release). These restrictions are expected to be addressed/removed in future software releases as ambient graduates to beta and eventually General Availability. Note that if a version earlier than 1.21.0 is used, there will be additional restrictions not listed here. 

Reviewer: Just use the alpha warning boilerplate here.

```suggestion
{{< boilerplate ambient-alpha-warning >}}

Since the ambient functionality is currently at an alpha release level, the following is a list of feature restrictions or caveats in ambient alpha. These restrictions are expected to be addressed/removed in future software releases as ambient graduates to beta and eventually General Availability. 
```

> 1. **Kubernetes (K8s) only:** Istio in ambient mode is currently only supported for deployment on Kubernetes clusters. Deployment on non-Kubernetes endpoints such as virtual machines is not currently supported.

Reviewer 2: ```suggestion
1. **Kubernetes (K8s) only:** Istio ambient mode is currently only supported for deployment on Kubernetes clusters. Deployment on non-Kubernetes endpoints such as virtual machines is not currently supported.
```

1. **No Istio multi-cluster support:** Only single cluster deployments are currently supported for Istio ambient mode.

1. **TCP/IPv4 only:** In the current release, TCP over IPv4 is the only protocol supported for transport on an Istio secure overlay tunnel (this includes protocols such as HTTP that run between application layer endpoints on top of the TCP/ IPv4 connection).

> 1. **No dynamic switching to ambient mode:** ambient mode can only be enabled on a new Istio mesh control plane that is deployed using ambient profile or ambient helm configuration. An existing Istio mesh deployed using a pre-ambient profile for instance can not be dynamically switched to also enable ambient mode operation.

Reviewer: ```suggestion
1. **Converting existing Istio deployments to ambient mode:** ambient mode can only be enabled on a new Istio mesh control plane that is deployed using ambient profile or ambient helm configuration. An existing Istio mesh deployed using a pre-ambient profile for instance can not be dynamically switched to also enable ambient mode operation.
```

1. **Cannot transparently convert existing Istio deployments to ambient mode:** Ambient mode can only be enabled on a new Istio mesh control plane that is deployed using the ambient `istioctl` profile or Helm configuration. An existing Istio mesh deployed using a sidecar profile cannot currently be dynamically switched to enable ambient mode.

1. **Restrictions with Istio `PeerAuthentication`:** as of the time of writing, the `PeerAuthentication` resource is not supported by all components (i.e. waypoint proxies) in Istio ambient mode. Hence it is recommended to only use the `STRICT` mTLS mode currently. Like many of the other alpha stage caveats, this shall be addressed as the feature moves toward beta status.

1. **istioctl CLI gaps:** There may be some minor functional gaps in areas such as Istio CLI output displays when it comes to displaying or monitoring Istio's ambient mode related information. These will be addressed as the feature matures.

### Environment used for this guide

The examples in this guide used a deployment of Istio version `1.21.0` on a `kind` cluster of version `0.20.0` running Kubernetes version `1.27.3`.

> 

Reviewer: ```suggestion
```

## Current Caveats {#caveats}

{{< boilerplate ambient-alpha-warning >}}

The following is a list of feature restrictions or caveats in ambient mode alpha. These restrictions are planned to be addressed or removed in future releases.

1. **Kubernetes only:** Istio in ambient mode is currently only supported for deployment on Kubernetes clusters. Deployment on non-Kubernetes endpoints such as virtual machines is not currently supported.

1. **No Istio multi-cluster support:** Only single cluster deployments are currently supported for Istio ambient mode.

1. **TCP/IPv4 only:** In the current release, TCP over IPv4 is the only protocol supported for transport on an Istio secure overlay tunnel (this includes protocols such as HTTP that run between application layer endpoints on top of the TCP/ IPv4 connection).

> 1. **Cannot transparently convert existing Istio deployments to ambient mode:** Ambient mode can only be enabled on a new Istio mesh control plane that is deployed using the ambient `istioctl` profile or Helm configuration. An existing Istio mesh deployed using a sidecar profile cannot be dynamically switched to enable ambient mode.

Reviewer 3: maybe add "at the moment" at the end? This is just a limitation for now, but not permanent imho.

> 1. **Cannot transparently convert existing Istio deployments to ambient mode:** Ambient mode can only be enabled on a new Istio mesh control plane that is deployed using the ambient `istioctl` profile or Helm configuration. An existing Istio mesh deployed using a sidecar profile cannot be dynamically switched to enable ambient mode.

Reviewer: ```suggestion
1. **Cannot transparently convert existing Istio deployments to ambient mode:** Ambient mode can only be enabled on a new Istio mesh control plane that is deployed using the ambient `istioctl` profile or Helm configuration. An existing Istio mesh deployed using a sidecar profile cannot be dynamically switched to enable ambient mode at the moment.
```

1. **Restrictions with Istio `PeerAuthentication`:** as of the time of writing, the `PeerAuthentication` resource is not supported by all components (i.e. waypoint proxies) in Istio ambient mode. Hence it is recommended to only use the `STRICT` mTLS mode currently. Like many of the other alpha stage caveats, this shall be addressed as the feature moves toward beta status.

1. **istioctl CLI gaps:** There may be some minor functional gaps in areas such as Istio CLI output displays when it comes to displaying or monitoring Istio's ambient mode related information. These will be addressed as the feature matures.

### Environment used for this guide

The examples in this guide used a deployment of Istio version `1.21.0` on a `kind` cluster of version `0.20.0` running Kubernetes version `1.27.3`.

> As ambient mode is not yet stable, the Istio authors request that you only test and file feedback on the most recent release.

Reviewer 3: suggest remove this as you have ambient stable warning earlier already.

> As ambient mode is not yet stable, the Istio authors request that you only test and file feedback on the most recent release.

Reviewer: ```suggestion
```

### Environment used for this guide

The examples in this guide used a deployment of Istio version `1.21.0` on a `kind` cluster of version `0.20.0` running Kubernetes version `1.27.3`.

> The recommended minimum Istio version needed for ambient functions is 1.21.0 and the minimum Kubernetes version needed is `1.24.0`. The examples below require a cluster with more than 1 worker node in order to explain how cross-node traffic operates. Refer to the [installation user guide](/docs/ops/ambient/install/) or [getting started guide](/docs/ops/ambient/getting-started/) for information on installing Istio in ambient mode on a Kubernetes cluster.

Reviewer: ```suggestion
As Istio Ambient is not yet stable, the most recent Istio release is the only recommended version for ambient use.

The examples below require a cluster with more than 1 worker node in order to explain how cross-node traffic operates. Refer to the [installation user guide](/docs/ops/ambient/install/) or [getting started guide](/docs/ops/ambient/getting-started/) for information on installing Istio in ambient mode on a Kubernetes cluster.
```

The figure shows that pods C1 and C2 running on node W1 connect with pod S1 running on node W2 and their TCP traffic is tunneled through HBONE tunnel instances that have been created by the ztunnel proxy pods of each node. Mutual TLS (mTLS) is used for encryption as well as mutual authentication of traffic being tunneled. SPIFFE identities are used to identify the workloads on each side of the connection. The term `HBONE` (for HTTP Based Overlay Network Encapsulation) is used in Istio ambient to refer to a technique for transparently and securely tunneling TCP packets encapsulated within HTTPS packets. For more details on the datapath including HBONE and the traffic redirection details, refer to the accompanying document [ztunnel datapath and inpod redirection](/docs/ops/ambient/usage/inpod). 

{{< tip >}}

> Note: Although the figure shows the HBONE tunnels to be between the two ztunnel proxies, in the current implementation with the inpod redirection model (from istio 1.21.0 onwards), the tunnels are in fact between the source and destination pods and traffic is HBONE encapsulated and encrypted in the network namespace of the source pod itself and eventually decapsulated and decrypted in the network namespace of the destination pod on the destination worker node. The ztunnel proxy still logically handles both the control plane and data plane needed for HBONE transport, however it is able to do that from inside the network namespaces of the source and destination pods. This is explained in more detail in the document [ztunnel datapath and inpod redirection](/docs/ops/ambient/usage/inpod).

Reviewer: nit: should we update the diagrams and remove this explainer, or keep it? 

I am actually not sure.

Reviewer 4: update diagram please

### Note on HBONE {#hbonesection}

HBONE (HTTP Based Overlay Network Encapsulation) is an Istio-specific term. It refers to the use of standard HTTP tunneling via the [HTTP CONNECT](https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/CONNECT) method to transparently tunnel application packets/ byte streams. In its current implementation within Istio, it transports TCP packets by tunneling these transparently using the HTTP CONNECT method, uses [HTTP/2](https://httpwg.org/specs/rfc7540.html), with encryption and mutual authentication provided by [mutual TLS](https://www.cloudflare.com/learning/access-management/what-is-mutual-tls/) and the HBONE tunnel itself runs on TCP port 15008. The overall HBONE packet format from IP layer onwards is depicted in the following figure.

{{< image width="100%"
link="hbone-packet.png"
caption="HBONE L3 packet format"
>}}

> Additional use cases of HBONE and HTTP tunneling (such as support for IPv6 and UDP packets) will be investigated in future as ambient mode evolves.

Reviewer 5: ```suggestion
Additional use cases of HBONE and HTTP tunneling (such as support for IPv6 and UDP packets) will be investigated in the future as ambient mode evolves.
```
